### ISSUE TITLE
*Animators implementation

### ISSUE THREAD

**Issue by stephenmdangelo** [qmlweb/qmlweb#382]

**Comment by ChALkeR** [comment]
I'm working on this.

**Comment by stephenmdangelo** [comment]
@ChALkeR I converted this from a Project card, as I believe it's the best place to comment on the *Animators implementation.  Feel free to delete if not :)

I'd love to see this implemented via CSS animations where possible for the smoothest user experience.  Since Animators don't have the requirement of the property being updated while the animation is running.

I've toyed around with using CSS transitions for some Behaviors, but you do end up running into issues like the fact that you can only apply a transition to an ENTIRE CSS transform, and not just the scale or translate portion like QML allows you to (without going and creating multiple DOM layers, one for each transform type).

**Comment by ChALkeR** [comment]
@stephenmdangelo  

> I'd love to see this implemented via CSS animations where possible for the smoothest user experience. Since Animators don't have the requirement of the property being updated while the animation is running.

Yes, I'm working on `@keyframes`-based implementation. Unfortunately, we can't use [Web Animations API](http://caniuse.com/web-animation) at this point, so that requires dynamically creating CSS code (which I already implemented, just the glueing code is not complete yet).

> you can only apply a transition to an ENTIRE CSS transform

I suppose this should work with adding multiple animations. I want to implement this without transitions first, just using the css animations instead, then take a look if this could be optimized with transitions.

See http://codepen.io/anon/pen/PWopjM for an example of css animations (this is how it would be done here). Note that there is also an `animationend` DOM event, that will allow us to better synchronize the css animations with QML property updates.

**Comment by ChALkeR** [comment]
Basically, I want to ignore css transitions for now and implement this like QML Animations, but using `@keyframes` CSS animations instead of a property-updating ticker. That means that Animators would be instantiated programmatically from the js code now, by adding an `animation` css property to an element, not automatically via `transition`s.

**Comment by ChALkeR** [comment]
I will take a closer look at CSS transforms now, give me a moment =).

**Comment by ChALkeR** [comment]
Update: yes, it looks that animations with CSS transforms are indeed conflicting with each others: http://codepen.io/anon/pen/mRdWjN.

**Comment by stephenmdangelo** [comment]
>  That means that Animators would be instantiated programmatically from the js code now, by adding an animation css property to an element, not automatically via transitions.

Oh yes, I agree not to use `transitions` for `Animators`. I only mentioned it to bring up the issues I had with testing smoothing out `Behavior` animations, since CSS `transitions` and QML `Behaviors` appear to be analogous concepts.

**Comment by ChALkeR** [comment]
@stephenmdangelo About the parallel animators conflicts — I suppose we can ignore that for now and implement a basic approach, and then decide what to do with parallel animators.

Note: parallel *Animators are [also broken in Qt QML](https://bugreports.qt.io/browse/QTBUG-42241).
